### PR TITLE
Authenticate to the Inbox using GitHub OAuth

### LINKED ISSUE #3: Oauth Login
Currently you have to make a personal access token and run the app yourself, we can [omniauth-github](https://github.com/intridea/omniauth-github) to enable people to login via a hosted version without having to run the app themselves.

### PR DESCRIPTION
**Pull request by tarebyte**
Closes #3 

Instead of asking a user to add their personal access token to the `.env` file, we can make an OAuth application and let the user authenticate against GitHub.

This isn't quite ready to go yet, it needs more tests for a starter. But I thought I'd push it up to see if anyone had initial feedback.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -9,6 +9,8 @@ gem 'uglifier'
 gem 'jquery-rails'
 gem 'turbolinks'
 gem 'octokit'
+gem 'omniauth'
```
Comment (andrew) [inline Gemfile]: This is a dependency of `omniauth-github`, I think we can leave it out

File: app/controllers/notifications_controller.rb [hunk 2]
```diff
@@ -13,17 +13,19 @@ def index
...
     scope = scope.type(params[:type]) if params[:type].present?
     scope = scope.status(params[:status]) if params[:status].present?
+    scope = scope.owned_by(current_user)
     @notifications = scope.newest
   end
 
   def archive
-    notification = Notification.find(params[:id])
-    notification.update_attributes(archived: true)
+    notification = Notification.find_by(id: params[:id], user: current_user)
```
Comment (andrew) [inline app/controllers/notifications_controller.rb]: we can do `notification = current_user.notifications.find(params[:id])` here

File: app/controllers/notifications_controller.rb [hunk 3]
```diff
@@ -13,17 +13,19 @@ def index
...
+    notification.update_attributes(archived: true) if notification.present?
```
Comment (andrew) [inline app/controllers/notifications_controller.rb]: If we use `find(params[:id])` in the line above then rails will raise an `NotFound` exception here if it's missing so no need to check for presence.

File: app/models/notification.rb [hunk 4]
```diff
@@ -17,13 +22,13 @@ def repo_url
     "https://github.com/#{repository_full_name}"
   end
 
-  def self.download
-    client = Octokit::Client.new(access_token: ENV['GITHUB_TOKEN'], auto_paginate: true)
+  def self.download(user)
+    client = Octokit::Client.new(access_token: user.access_token, auto_paginate: true)
```
Comment (andrew) [inline app/models/notification.rb]: Can we move the client onto a method on user called `github_client`?

**Comment by andrew** [comment]
Lovely work, I left a few comments but it's looking good! We'll need to update the readme with instructions on making your own oauth app on github

**Comment by andrew** [comment]
Also conflict in schema.rb after I merged @xadamy's latest pr :feelsgood:

File: app.json [hunk 5]
```diff
@@ -6,8 +6,12 @@
     "postdeploy": "bundle exec rake db:migrate"
   },
   "env": {
-    "GITHUB_TOKEN": {
-      "description": "Your personal GitHub Access Token. It must have access to the 'notification' scope.",
+    "GITHUB_CLIENT_ID": {
+      "description": "The GitHub Application Client ID",
+      "required": true
+    }
```
Comment (jeroenvisser101) [inline app.json]: A comma is missing here.

File: Gemfile (hunk 1 above)
Comment (tarebyte) [inline Gemfile]: Resolved by ffbd756

File: app/controllers/notifications_controller.rb (hunk 2 above)
Comment (tarebyte) [inline app/controllers/notifications_controller.rb]: Resolved by 0edb8bd

**Comment by tarebyte** [comment]
@andrew I think this is ready for another 👀

Thanks for the previous feedback ✨

**Comment by andrew** [comment]
Would be good to have some kind of introductory homepage for logged out users rather than just punting them straight to the GitHub oauth page, can be super simple for now and we can improve it later

**Comment by andrew** [comment]
Small conflict in the layout after the [rename](https://github.com/andrew/octobox/pull/38) to Octobox.

**Comment by andrew** [comment]
Believe it or not, my first time using PR reviews, I've been working alone for so long 😆

File: app/controllers/notifications_controller.rb [hunk 6]
```diff
@@ -1,26 +1,32 @@
...
+
+    scope = scope.repo(params[:repo])     if params[:repo].present?
     scope = scope.reason(params[:reason]) if params[:reason].present?
-    scope = scope.type(params[:type]) if params[:type].present?
+    scope = scope.type(params[:type])     if params[:type].present?
     scope = scope.status(params[:status]) if params[:status].present?
-    scope = scope.starred if params[:starred].present?
+    scope = scope.starred                 if params[:starred].present?
+
+    scope = scope.owned_by(current_user)
```
Comment (WillAbides) [inline app/controllers/notifications_controller.rb]: This should go above the count assignments so that it only shows counts for the current user

Comment (andrew) [inline app/controllers/notifications_controller.rb]: 👍 Infact we can change the scope at the top of the method from `Notification` to `current_user.notifications`

File: app/models/user.rb [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+  def self.find_by_auth_hash(auth_hash)
+    User.find_by(github_id: auth_hash['uid'])
+  end
+
+  def assign_from_auth_hash(auth_hash)
+    github_attributes = {
+      github_id: auth_hash['uid'],
+      access_token: auth_hash.dig('credentials', 'token')
+    }
```
Comment (tarebyte) [inline app/models/user.rb]: We can, I always hesitate about storing the `login` since it can change. But I'll go ahead update it.

Comment (andrew) [inline app/models/user.rb]: I figure we'd only use it as a display in the UI along with the avatar of the current user, makes it much easier to link to a GitHub profile without making a http request too

File: app/controllers/notifications_controller.rb (hunk 6 above)
Comment (tarebyte) [inline app/controllers/notifications_controller.rb]: Resolved by 0bf79a9

File: config/initializers/omniauth.rb [hunk 8]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+Rails.application.config.middleware.use OmniAuth::Builder do
+  provider :github,
+           Rails.application.secrets.github_client_id,
+           Rails.application.secrets.github_client_secret,
+           scope: 'notifications'
```
Comment (tarebyte) [inline config/initializers/omniauth.rb]: Resolved by c0643ab

File: app/models/user.rb (hunk 7 above)
Comment (tarebyte) [inline app/models/user.rb]: Resolved by ae8ccc2

**Comment by tarebyte** [comment]
This is the homepage right now, super basic.

screen shot 2016-12-18 at 21 30 49

File: README.md [hunk 9]
```diff
@@ -76,10 +76,11 @@ bundle exec rake db:create:all
 bundle exec rake db:migrate
 ```
 
-Now go and create a [personal access token](https://github.com/settings/tokens) on GitHub with the `notifications` scope enabled and add it to `.env`:
+Now go and register a new [GitHub OAuth Application](https://github.com/settings/applications/new) then add the following to your `.env`:
```
Comment (tarebyte) [inline README.md]: Let me know what you think of the changes in b69fa85

**Comment by tarebyte** [comment]
This is what the auth flow looks like.

auth

**Comment by tarebyte** [comment]
@andrew I think I've hit all of your feedback, let me know what you think!

**Comment by WillAbides** [comment]
Does this still allow the use of a personal access token?  I'd hate to not be able to use octobox for notifications from orgs that won't approve access from third party apps.

**Comment by tarebyte** [comment]
> Does this still allow the use of a personal access token? I'd hate to not be able to use octobox for notifications from orgs that won't approve access from third party apps.

@WillAbides right now it doesn't, however I'd like to follow up this PR with another that would let a user set their token after signing in. Cause I feel your pain :octocat:

**Comment by andrew** [comment]
Amazing work @tarebyte 🎉 

Adding back in the option for using a personal access token would be 👌 @WillAbides can you open an issue for it?

**Pull request merged**